File: app/services/statistics.rb
class Statistics
  def initialize(params)
    @params = params
    @performed_scope = {}
    perform_scope
  end

  # TODO
  # private

  def perform_scope
    @scopes = @params[:scopes]
    return unless @scopes.present?

    @scopes = @scopes.split(',')
    @scopes.each do |scope|
      @performed_scope[scope] = send(scope.to_sym) if respond_to?(scope.to_sym)
    end
    @performed_scope
  end

  # Stats
  def general_stat
    {
      active_days: stars.uniq(&:due_date).count
    }
  end

  def tasks_stat
    {
      all:        tasks.count,
      all_active: tasks.created.count
    }
  end

  def goals_stat
    {
      all: tasks.count,
      current: {
        id: current_active_goal.id,
        number_of_stars: current_active_goal.number_of_stars,
        completed_stars: stars.where('due_date BETWEEN ? AND ?',
                        current_active_goal.start_date, current_active_goal.due_date).count
      }
    }
  end

  def stars_stat
    {
      all:           stars.count,
      current_year:  stars_count_between_dates('year'),
      current_month: stars_count_between_dates('month'),
      current_week:  stars_count_between_dates('week'),
      current_day:   stars_count_between_dates('day')
    }
  end

  # Charts
  def heatmap_chart
    stars.group_by(&:due_date).transform_values(&:count).map do |key, value|
      { date: key, count: value }
    end
  end

  def treemap_chart
    [{
      data: stars.preload(:task).group_by(&:task).transform_values(&:count).map do |key, value|
        { x: key.title, y: value }
      end
    }]
  end

  private

  # Varilables (lazy loading)
  def stars
    @stars ||= Star.all
  end

  def tasks
    @tasks ||= Task.all
  end

  def goals
    @goals ||= Goal.all
  end

  def current_active_goal
    @current_active_goal ||= goals.created.first
  end

  # Methods
  def stars_count_between_dates(period)
    @stars.where('due_date BETWEEN ? AND ?',
                  DateTime.now.send("beginning_of_#{period}"), DateTime.now.send("end_of_#{period}")).count
  end
end

File: app/javascript/app/store/modules/statistics.js
export default {
  state: {
    general: {
      active_days: 0
    },
    tasks: {
      all: 0,
      all_active: 0
    },
    goals: {
      all: 0,
      current: {
        id: 0,
        number_of_stars: 0,
        completed_stars: 0
      }
    },
    stars: {
      all: 0,
      current_year: 0,
      current_month: 0,
      current_week: 0,
      current_day: 0
    },
    heatmap: {},
    treemap: {}
  },
  mutations: {
    updateTasksStatistics(state, values) {
      state.general.active_days = values.active_days
    },
    updateTasksStatistics(state, values) {
      state.tasks.all        = values.all,
      state.tasks.all_active = values.all_active
    },
    updateGoalsStatistics(state, values) {
      state.goals.all                     = values.all,
      state.tasks.current.number_of_stars = values.number_of_stars
    },
    updateStarsStatistics(state, values) {
      state.stars.all           = values.all,
      state.stars.current_year  = values.current_year,
      state.stars.current_month = values.current_month,
      state.stars.current_week  = values.current_week,
      state.stars.current_day   = values.current_day
      state.stars.current_goal  = values.current_goal
    },
    updateHeatmapStatistics(state, values) {
      state.heatmap = values
    },
  },
  actions:   {}
}